GrandCentralBoard/Widgets/Bonus/Source/BonusSource.swift

```diff
@@ -0,0 +1,129 @@
+
+final class BonusSource : Asynchronous {
+    
+    typealias ResultType = Result<[Person]>
+    typealias UpdateResultType = Result<Updates>
+    typealias ImageResultType = Result<UIImage>
+
+    let sourceType: SourceType = .Cumulative
+    let interval: NSTimeInterval = 5
+    
+    var mapping: Mapping?
+    private var people: [String:Person] = [:]
+    private let dataDownloader: DataDownloading
+    private let settings: BonusWidgetSettings
+    
+    init(settings: BonusWidgetSettings, dataDownloader: DataDownloading) {
+        self.dataDownloader = dataDownloader
+        self.settings = settings
+    }
+    
+    func read(closure: (ResultType) -> Void) {
+        
+        fetchBonusUpdate { result in
+            switch result {
+                case .Success(let update):
+                    let updates = update.all.sort({ return $0.date.isLessThanDate($1.date) })
+                    self.updatePeople(updates)
+                    self.fetchImages(closure)
+                case .Failure(let error):
+                    closure(.Failure(error))
+            }
+        }
+    }
+    
+    private func updatePeople(updates: [Update]) {
+        updates.forEach { update in
+            guard let person = people[update.name] else {
+                people[update.name] = Person.personFromUpdate(update, imageUrl: mapping?.data[update.name])
+                return
+            }
+            
+            if let url = mapping?.data[update.name] {
+                people[update.name] = person.copyPersonWithImage(BubbleImage(url: url))
+            }
+            
+            guard person.lastUpdate.isLessThanDate(update.date) else { return }
+            people[update.name] = person.copyByUpdating(update, imageUrl: mapping?.data[update.name])
+        }
+    }
+    
+    // MARK: Fetching data from bonus.ly
+    
+    private func fetchBonusUpdate(completion: (UpdateResultType) -> Void) {
+        dataDownloader.downloadDataAtPath(settings.bonuslyPath) { result in
+            switch result {
+                case .Success(let data):
+                    do {
+                        let updates = try Updates.updatesFromData(data)
+                        completion(.Success(updates))
+                    } catch (let error) {
+                        completion(.Failure(error))
+                    }
+                case .Failure(let error):
+                    completion(.Failure(error))
+            }
+        }
+    }
+    
+    // MARK: Fetching images
+    
+    private func fetchImages(closure: (ResultType) -> Void) {
+        let imagesToDownloadCount = people.filter({ (name, person) -> Bool in
+            if let _ = person.bubbleImage.url where person.bubbleImage.remoteImage == nil {
+                return true
+            }
+            return false
+        }).count
+        
+        guard imagesToDownloadCount > 0 else { return closure(.Success(Array(people.values))) }
+        let downloadsCounter = AtomicCounter()
```

Comment:
Can I recommend using an dispatch group instead?

But I'm almost sure you don't need any of it since `Alamofire` calls completion block on the same thread (main) and this is enough synchronization to decrement or even pop something from an array.